### ISSUE TITLE
[UT] test_emulate_precision_casts_preserves_explicit_precision_cast_float16 fails

### ISSUE THREAD

**Issue by ZhaoqiongZ** [intel/torch-xpu-ops#4358]
## Summary

`test/inductor/test_cuda_repro.py::CudaReproTests::test_emulate_precision_casts_preserves_explicit_precision_cast_float16`
fails on XPU. The `float16` parametrization diverges from eager when
`torch._inductor.config.emulate_precision_casts = True`, because the explicit
`fp32 -> fp16 -> fp32` precision barrier is silently eliminated by the
triton-xpu backend.

This issue is the canonical tracker referenced by the source-level skip
added in pytorch/pytorch (skip PR: to be filled in after opening). The skip
targets only the `float16` parametrization; the `bfloat16` variant is not
affected and continues to run (see "Why only float16" below).
Do **not** close this issue until the triton-xpu fix has landed in the pinned
wheel and the skip has been removed from the test source.

Related upstream: pytorch/pytorch#190157 (dynamic DISABLED, superseded by the
source-level skip once the PR above lands).

## Environment

- torch: `2.14.0.dev20260713+xpu`
- triton-xpu: `3.7.2+git5fcc14d9`
- XPU: Intel(R) Data Center GPU Max 1100 (also expected on other Intel GPUs)

## Reproducer (matches upstream UT shape)

```python
import torch
import torch._inductor.config as ind
from torch._inductor.utils import run_and_get_code

torch.manual_seed(0); torch.xpu.manual_seed_all(0)
x = torch.randn(4, 32, 32, device="xpu", dtype=torch.float32)
w = torch.randn(32, 32, device="xpu", dtype=torch.float32)

def fn(x, w):
    x = torch.matmul(x, w)
    x = x.to(torch.float16).float()
    x = x * torch.sigmoid(x)
    return x.sum(dim=1)

with ind.patch(emulate_precision_casts=True):
    opt_fn = torch.compile(fn, backend="inductor", fullgraph=True)
    expected = fn(x, w)
    actual, (code,) = run_and_get_code(opt_fn, x.clone(), w.clone())

print("max abs diff:", (expected - actual).abs().max().item())
# XPU:  0.017578125  -> assertEqual FAILS
# CUDA: 0.0
```

## Isolation results (rules out PyTorch / Inductor as the offender)

| variant                                    | max abs diff |
|--------------------------------------------|--------------|
| matmul only                                | 0.0          |
| matmul + sigmoid + sum, no fp16 cast       | 1.5e-5       |
| `.to(fp16).float()` alone (cast round-trip) | **7.7e-3**  |
| `.to(bf16).float()` alone (cast round-trip) | **0.0**     |
| full UT (fp16 variant)                     | **0.018**    |
| full UT (bf16 variant)                     | 7.6e-6 (from downstream sigmoid / `sum(dim=1)` reduction order, not the cast; PASS) |

The error is produced entirely by the `.to(fp16).float()` step; all other
operations are numerically exact.

## Root cause (in triton-xpu, not in PyTorch)

Inductor codegen is correct: the generated Triton source contains the
expected barrier `tmp1 = tmp0.to(tl.float16); tmp2 = tmp1.to(tl.float32)`.
TTIR / TTGIR / LLIR all preserve `arith.truncf` + `arith.extf` (LLVM
`fptrunc float, half` + `fpext half, float`). `-cl-opt-disable`
(`TRITON_INTEL_DISABLE_IGC_OPT=1`) does **not** change the numerical result,
so it is not a user-tunable IGC optimization either.

The bug is in
`intel-xpu-backend-for-triton/third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp`
in `TruncFOpConversion::createDestOps` (currently around line 1442):
`arith.truncf f32 -> f16` is lowered to a plain `LLVM::FPTruncOp`, which the
downstream SPIR-V / IGC backend then folds against the following `fpext` as
identity.

The same file already provides a correct lowering for `tt.fp_to_fp` in
`FpToFpOpConversion::convertFp32ToFp16` (around line 1097), which uses
`LLVM::ConstrainedFPTruncIntr` (constrained intrinsic with explicit rounding
mode and fp exception behavior). Constrained intrinsics carry side-effect
metadata that prevents the folding. The two paths must be aligned.

## Why only float16 (bf16 is not affected)

The `bfloat16` path in `TruncFOpConversion::createDestOps` goes through
`intel::convertFp32ToBf16` (either the `__spirv_ConvertFToBF16INTEL`
intrinsic or a hand-written RTNE bit-manipulation sequence in
`BF16Casts.cpp`), which never emits an LLVM `fptrunc`. The
"unconstrained `fptrunc`+`fpext` folded away as identity" root cause
simply does not apply to bf16. Isolating the cast alone confirms this on
XPU today: `matmul(x, w).to(bf16).float()` is bit-exact against eager
(diff = 0.0).

The ~7.6e-6 residual in the full bf16 UT comes from downstream ops
(sigmoid / `sum(dim=1)` reduction order), i.e. ordinary floating-point
reordering variance between compiled and eager execution, unrelated to
`emulate_precision_casts`. `assertEqual`'s default fp32 tolerance
correctly absorbs it, and bf16's pass is a genuine pass, not a false pass.

Consequence for the upstream fix: touch **only** the `f32 -> f16` branch
of `TruncFOpConversion`; do not modify the bf16 path (it is currently
correct).

## Fix location and approach (upstream: intel-xpu-backend-for-triton)

Route the `f32 -> f16` case of `arith.truncf` through the same
`ConstrainedFPTruncIntr` used by `tt.fp_to_fp`. Minimal patch in
`ElementwiseOpToLLVM.cpp::TruncFOpConversion::createDestOps`:

```cpp
auto inElemTy = getElementType(op.getIn());
if (inElemTy.isF32() && outElemTy.isF16()) {
  return {FpToFpOpConversion::convertFp32ToFp16(
      loc, rewriter, operands[0][0], RoundingMode::RTNE)};
}
return {LLVM::FPTruncOp::create(rewriter, loc, elemTy, operands[0][0])};
```

Suggested tests to add in intel-xpu-backend-for-triton:
- `test/TritonIntelGPU/`: lit test asserting `arith.truncf f32 -> f16`
  lowers to `llvm.experimental.constrained.fptrunc`.
- `python/test/unit/`: numerical e2e test — a hand-written Triton kernel
  running `x.to(tl.float16).to(tl.float32)` must actually round-trip through
  fp16 (compare against a reference computed on CPU with numpy fp16).

## Dependency chain / closure conditions

- [x] Upstream bug filed: intel/intel-xpu-backend-for-triton#7491
- [ ] Fix PR merged in `intel/intel-xpu-backend-for-triton` (link when opened).
- [ ] Released triton-xpu build / commit containing the fix.
- [ ] PyTorch triton-xpu commit pin bumped in
      `.ci/docker/ci_commit_pins/` (or equivalent) to a build containing the fix.
- [ ] XPU CI re-runs the UT and it is green.
- [ ] Skip removed from `test/inductor/test_cuda_repro.py`
      (the `raise unittest.SkipTest(...)` guard inside
      `test_emulate_precision_casts_preserves_explicit_precision_cast`).
- [ ] pytorch/pytorch#190157 closed (dynamic DISABLED marker superseded).

**Only after all boxes above are checked may this issue be closed.**
Closing this issue is the signal that the PyTorch-side skip decorator may
be removed.

**Comment by ZhaoqiongZ** [comment]
Skip PR opened: pytorch/pytorch#190163 (source-level skip inside `test_emulate_precision_casts_preserves_explicit_precision_cast`, targets `float16` parametrization only; `bfloat16` continues to run).

**Comment by ZhaoqiongZ** [comment]
@torchxpubot triage

**Comment by torchxpubot** [comment]
## Issue Triage

| Field | Value |
|-------|-------|
| Reproduction missing | no |
| Scope | pytorch |
| Dependencies | triton, driver |
| Handling | needs-human |

**Reason:** runtime dependency requires human triage: triton, driver

*Automated by issue-triage.*

**Comment by ZhaoqiongZ** [comment]
@torchxpubot fix

**Comment by torchxpubot** [comment]
Starting automated fix on this issue. Follow progress in the [workflow run](https://github.com/intel/torch-xpu-ops/actions/runs/32444424657).

**Comment by torchxpubot** [comment]
@chuanqi129 `@torchxpubot fix` failed: the agent returned an error. This is usually a Bedrock or model failure; check AWS token validity.

See the [`fix` job](https://github.com/intel/torch-xpu-ops/actions/runs/32444424657/job/96661261878) for details.
